File: controllability_observability/connectedsys.py
# controllability_observability/connectedsys.py
import control
import numpy as np

A1 = [[0, 1], [-2, -3]]
B1 = [[0], [1]]
C1 = [[2, 1]]
sys1 = control.ss(A1, B1, C1, 0)
P1 = control.ctrb(A1,B1)
print(np.linalg.matrix_rank(P1))
Q1 = control.obsv(A1,C1)
print(np.linalg.matrix_rank(Q1))

A2 = [[-1, 0], [0, -3]]
B2 = [[1], [1]]
C2 = [[1, 1]]
sys2 = control.ss(A2, B2, C2, 0)
P2 = control.ctrb(A2,B2)
print(np.linalg.matrix_rank(P2))
Q2 = control.obsv(A2,C2)
print(np.linalg.matrix_rank(Q2))

sys_s = control.series(sys1, sys2)
sys_p = control.parallel(sys1, sys2)

Ps = control.ctrb(sys_s.A, sys_s.B)
print(np.linalg.matrix_rank(Ps))
Qs = control.obsv(sys_s.A, sys_s.C)
print(np.linalg.matrix_rank(Qs))

Pp = control.ctrb(sys_p.A, sys_p.B)
print(np.linalg.matrix_rank(Pp))
Qp = control.obsv(sys_p.A, sys_p.C)
print(np.linalg.matrix_rank(Qp))

File: controllability_observability/connectedsys.m
% controllability_observability/connectedsys.m
A1 = [ 0 1; -2 -3 ]; B1 = [ 0; 1 ]; C1 = [ 2 1 ];
sys1 = ss(A1, B1, C1, 0)
P1 = ctrb(sys1)
rank(P1)
Q1 = obsv(sys1)
rank(Q1)

A2 = [ -1 0; 0 -3 ]; B2 = [ 1; 1 ]; C2 = [ 1 1 ];
sys2 = ss(A2, B2, C2, 0)
P2 = ctrb(sys2)
rank(P2)
Q2 = obsv(sys2)
rank(Q2)

sys_s = series(sys1,sys2)
sys_p = parallel(sys1,sys2)
Ps = ctrb(sys_s)
rank(Ps)
Qs = obsv(sys_s)
rank(Qs)

Pp = ctrb(sys_p)
rank(Pp)
Qp = obsv(sys_p)
rank(Qp)
